fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadWriteNonCoreDataServiceImpl.java

```diff
@@ -214,6 +214,185 @@ public List<JsonObject> queryDataTable(String datatable, final String columnFilt
+        List<ColumnFilter> columnFilters = request.map(DatatableSearchRequest::getColumnFilters).orElse(Collections.emptyList());
+        String datatable = request.map(DatatableSearchRequest::getDatatable).orElse(null);
+        Pageable pageable = searchRequest.toPageable();
+
+        this.context.authenticatedUser().validateHasDatatableReadPermission(datatable);
+
+        List<JsonObject> results = new ArrayList<>();
+        AtomicInteger totalElements = new AtomicInteger(0);
+        if (CollectionUtils.isNotEmpty(columnFilters) && CollectionUtils.isNotEmpty(resultColumns)) {
+
+            String resultColumn = resultColumns.stream().collect(Collectors.joining(","));
+
+            validateSqlInjectionForDynamicQuery(columnFilters, datatable, resultColumn);
+
+            List<ResultsetColumnHeaderData> resultsetColumnHeaderData = genericDataService.fillResultsetColumnHeaders(datatable);
+
+            if (CollectionUtils.isNotEmpty(resultsetColumnHeaderData)) {
+
+                Map<String, String> dataTableColumnMap = resultsetColumnHeaderData.stream()
+                        .collect(Collectors.toMap(ResultsetColumnHeaderData::getColumnName, ResultsetColumnHeaderData::getColumnType));
+                columnFilters.forEach(column -> {
+                    validateRequestParamsSearch(column.getColumnName(), column.getColumnValue(), column.getColumnOperation(), resultColumn,
+                            dataTableColumnMap);
+                });
+
+                AtomicInteger count = new AtomicInteger(0);
+
+                // Attach the selection
+                StringBuilder sqlBuilder = new StringBuilder();
+                if (databaseTypeResolver.isPostgreSQL()) {
+                    sqlBuilder.append("SELECT ").append(escapeFieldNames(resultColumn)).append(" FROM ")
+                            .append(sqlGenerator.escape(datatable));
+                } else if (databaseTypeResolver.isMySQL()) {
+                    sqlBuilder.append("SELECT ").append(resultColumn).append(" FROM ").append(datatable);
+                } else {
+                    throw new IllegalStateException("Database type is not supported");
+                }
+
+                // Build base query with filters but without the selection
+                String baseFilterQuery = buildBaseQueryWithFilters(columnFilters, dataTableColumnMap, count);
+
+                // Attach the count selection
+                StringBuilder countQueryBuilder = new StringBuilder();
+                countQueryBuilder.append(" SELECT COUNT(*) FROM ").append(datatable);
+
+                // Attach the WHERE clause
+                if (baseFilterQuery.length() > 0) {
+                    sqlBuilder.append(" WHERE ").append(baseFilterQuery);
+                    countQueryBuilder.append(" WHERE ").append(baseFilterQuery);
+                }
+
+                // Attach the ORDER
+                attachOrdering(sqlBuilder, pageable.getSort());
```

Comment:
Please decide whether your methods are amending the `sqlBuilder` or returning a value and that value is appended to the `sqlBuilder` but do not mix them... it's more readable if this is consistent.
**Example:**
```String baseFilterQuery = buildBaseQueryWithFilters(columnFilters, dataTableColumnMap, count);
sqlBuilder.append(" WHERE ").append(baseFilterQuery);
```
vs
```
attachOrdering(sqlBuilder, pageable.getSort());
```